services/datalad/datalad_service/handlers/files.py

```diff
@@ -22,8 +22,8 @@ def on_get(self, req, resp, dataset, filename, snapshot='HEAD'):
             # If the file begins with an annex path, return that path
             if file_content[0:4096].find('.git/annex') != -1:
                 # Resolve absolute path for annex target
-                target_path = os.path.join(
-                    ds_path, os.path.dirname(filename), file_content)
+                target_path = os.path.normpath(os.path.join(
+                    ds_path, os.path.dirname(filename), file_content))
                 # Verify the annex path is within the dataset dir
                 if ds_path == os.path.commonpath((ds_path, target_path)):
```

Comment:
The test fails without this? I'm having trouble understanding how this changes the semantics at all. Because we get "file not found" and not "file not found in git tree", we're getting an `IOError` instead of failing this path check. We could plausibly be getting that on `git_show()` (L21), `open` (L29) or `fstat` (L31).

Is there a case where `open(path)` and `open(normpath(path))` should have any different result?

Reply:
There is a difference, the path referenced has been removed. That's why this works for paths that still exist in the draft. You can't traverse up the directory tree from a directory that can't be opened. Normpath gives us the path without the potentially missing components by converting git-annex's dataset-root/../../.git paths into dataset-root/.git paths that do exist.

Reply:
Oh, so the history part actually is a requirement then. Good thing my test makes more sense than my PR description.

Reply:
Ah, thank you. That's what I was missing. This LGTM, then.